pkg/controller/storageversionmigrator/storageversionmigrator.go

```diff
@@ -0,0 +1,289 @@
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, svmc.worker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (svmc *SVMController) worker(ctx context.Context) {
+	for svmc.processNext(ctx) {
+	}
+}
+
+func (svmc *SVMController) processNext(ctx context.Context) bool {
+	svmKey, quit := svmc.queue.Get()
+	if quit {
+		return false
+	}
+	defer svmc.queue.Done(svmKey)
+
+	key := svmKey.(string)
+	err := svmc.sync(ctx, key)
+	if err == nil {
+		svmc.queue.Forget(key)
+		return true
+	}
+
+	klog.FromContext(ctx).V(2).Info("Error syncing SVM resource, retrying", "svm", key, "err", err)
+	svmc.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (svmc *SVMController) sync(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	startTime := time.Now()
+
+	// SVM is a cluster scoped resource so we don't care about the namespace
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	svm, err := svmc.svmListers.Get(name)
+	if err != nil {
+		return err
+	}
+	// working with a copy to avoid race condition between this and resource version controller
+	toBeProcessedSVM := svm.DeepCopy()
+
+	if HasCondition(toBeProcessedSVM, svmv1alpha1.MigrationSucceeded) {
+		logger.V(4).Info("Migration has already succeeded, skipping", "svm", name)
+		return nil
+	}
+
+	if toBeProcessedSVM.Status.ResourceVersion == "" {
+		logger.V(4).Info("list resource version is empty, requeuing to attempt again", "svm", name)
+		svmc.queue.AddRateLimited(key)
+		return nil
+	}
+	gvr := getGVRFromResource(toBeProcessedSVM)
+
+	// wait for the dependency graph to be synchronized before migration to avoid a nil pointer.
+	// this prevents deadlocks in scenarios where available resources have changed, and one of the informer caches will never fill.
+	if !cache.WaitForNamedCacheSync(svmc.gcCollector.Name(), ctx.Done(), func() bool {
+		return svmc.gcCollector.IsSynced(logger)
+	}) {
+		return fmt.Errorf("dependency graph is not synced, requeuing to attempt again")
+	}
+
+	gcListResourceVersion, err := convertResourceVersionToInt(svmc.gcCollector.GetMonitorController(gvr).LastSyncResourceVersion())
+	if err != nil {
+		return err
+	}
+	listResourceVersion, err := convertResourceVersionToInt(toBeProcessedSVM.Status.ResourceVersion)
+	if err != nil {
+		return err
+	}
+
+	if gcListResourceVersion < listResourceVersion {
+		return fmt.Errorf("GC cache is not up to date, requeuing to attempt again. gcListResourceVersion: %d, listResourceVersion: %d", gcListResourceVersion, listResourceVersion)
+	}
+
+	toBeProcessedSVM, err = updateSVMStatus(
+		ctx,
+		svmc.kubeClient,
+		toBeProcessedSVM,
+		svmv1alpha1.MigrationRunning,
+		migrationRunningStatusReason,
+	)
+	if err != nil {
+		return err
+	}
+
+	// ToDo: implement a mechanism to resume migration from the last migrated resource in case of a failure
+	// process storage migration
+	for _, gvrKey := range svmc.gcCollector.GetMonitorStore(gvr).ListKeys() {
```

Comment:
looks like you're actually trying to access the GraphBuilder?  And probably a smaller interface off that really.

Reply:
Yup. Do I need to change anything here?

Reply:
> Yup. Do I need to change anything here?

Yes.  Create the interface you need, then add that interface to the controllerContext (not the entire garbage collector), then narrow the instance variable here to the interface you need.